packages/mdc-side-sheet/dismissible/foundation.js

```diff
@@ -0,0 +1,200 @@
+
+/**
+ * @extends {MDCFoundation<!MDCSideSheetAdapter>}
+ */
+class MDCDismissibleSideSheetFoundation extends MDCFoundation {
+  /** @return enum {string} */
+  static get strings() {
+    return strings;
+  }
+
+  /** @return enum {string} */
+  static get cssClasses() {
+    return cssClasses;
+  }
+
+  static get defaultAdapter() {
+    return /** @type {!MDCSideSheetAdapter} */ ({
+      addClass: (/* className: string */) => {},
+      removeClass: (/* className: string */) => {},
+      hasClass: (/* className: string */) => {},
+      elementHasClass: (/* element: !Element, className: string */) => {},
+      notifyClose: () => {},
+      notifyOpen: () => {},
+      saveFocus: () => {},
+      restoreFocus: () => {},
+      focusActiveNavigationItem: () => {},
+      trapFocus: () => {},
+      releaseFocus: () => {},
+    });
+  }
+
+  constructor(adapter) {
+    super(Object.assign(MDCDismissibleSideSheetFoundation.defaultAdapter, adapter));
+
+    /** @private {number} */
+    this.animationFrame_ = 0;
+
+    /** @private {number} */
+    this.animationTimer_ = 0;
+  }
+
+  destroy() {
+    if (this.animationFrame_) {
+      cancelAnimationFrame(this.animationFrame_);
+    }
+    if (this.animationTimer_) {
+      clearTimeout(this.animationTimer_);
+    }
+  }
+
+  /**
+   * Function to open the side-sheet.
+   */
+  open() {
+    if (this.isOpen() || this.isOpening() || this.isClosing()) {
+      return;
+    }
+
+    this.adapter_.addClass(cssClasses.OPEN);
+    this.adapter_.addClass(cssClasses.ANIMATE);
+
+    // Wait a frame once display is no longer "none", to establish basis for animation
+    this.runNextAnimationFrame_(() => {
+      this.adapter_.addClass(cssClasses.OPENING);
+    });
+
+    this.adapter_.saveFocus();
+  }
+
+  /**
+   * Function to close the side-sheet.
+   */
+  close() {
+    if (!this.isOpen() || this.isOpening() || this.isClosing()) {
+      return;
+    }
+
+    this.adapter_.addClass(cssClasses.CLOSING);
+  }
+
+  /**
+   * Extension point for when side-sheet finishes open animation.
+   * @protected
+   */
+  opened() {}
+
+  /**
+   * Extension point for when side-sheet finishes close animation.
+   * @protected
+   */
+  closed() {}
+
+  /**
+   * Returns true if side-sheet is in open state.
+   * @return {boolean}
+   */
+  isOpen() {
+    return this.adapter_.hasClass(cssClasses.OPEN);
+  }
+
+  /**
+   * Returns true if side-sheet is animating open.
+   * @return {boolean}
+   */
+  isOpening() {
+    return this.adapter_.hasClass(cssClasses.OPENING);
+  }
+
+  /**
+   * Returns true if side-sheet is animating closed.
+   * @return {boolean}
+   */
+  isClosing() {
+    return this.adapter_.hasClass(cssClasses.CLOSING);
+  }
+
+  /**
+   * Keydown handler to close side-sheet when key is escape.
+   * @param evt
+   */
+  handleKeydown(evt) {
+    const {keyCode, key} = evt;
+
+    const isEscape = key === 'Escape' || keyCode === 27;
+    if (isEscape) {
+      this.close();
```

Comment:
I think we'll need to parameterize this, because drawers should do this by default but side sheets should not (they may contain child controls that do something on escape).

We can enter a separate task for this if you would rather do it that way.

Reply:
Created #3938 for this 🙂